Refactoring Scad code

Neon22 opened the pull request:
Here I have made a start on refactoring.
I have adjusted only one file as a test to see if its useful to you.

- added a number of temp variables to hold repeatedly used variables.
- Also minor tab adjustments.

I would probably add a further PR which would add an internal module which is the screw (hole) and then refactor it so it was called for the left and right halves efficiently - while trying to maintain clarity.

I think clarity of intention is more important than making "perfect" code style.
But it would mean that changing screws in the future could be less prone to error.

I have made a large number of openscad projects on Thingiverse and use the customiser a lot also.
- https://www.thingiverse.com/neon22/designs
- I am also current maintainer of pyweaving.

Please let me know if this is useful and I will slowly continue adding PRs for your acceptance.

Neon22 committed:
first pass at refactoring

Here I have made a number of temp variables to hold repeatedly used variables.
Also minor tab adjustments.

Neon22 committed:
forgot to clean up after myself

A small adj to arg placement

Neon22 commented:
I foolishly adjust the tabs in the same commit so its much harder to see what I have done. Apologies.
Next time I would resubmit as several commits so its clearer the changes I am making

ScarlettSparks commented:
Thank you so much for this! I haven't had a chance to fully review all the details yet, but this looks like a sensible (and much-needed) refactoring approach. 

> I would probably add a further PR which would add an internal module which is the screw (hole) and then refactor it so it was called for the left and right halves efficiently - while trying to maintain clarity.

When it comes to this particular file it's worth noting that I'm hoping to do some re-designing to eliminate the needlebed screws as soon as I possibly can, so I hesitate to suggest pouring a lot of effort into related modules that might end up deprecated. With that said, if you still think that's an area you'd like to clean up, I'd appreciate seeing your approach!

Neon22 commented:
OK. Thanks. I will continue in a day or so. Cheers...

BTW - I have an ancient all-metal Corona which used metal sinkers at the end of each row.
I used to get stuff done through Ponoko but they changed their tack several years ago and I have not used them since.
But I was thinking about what would be involved to make metal parts like the Corona sinkers and looked them up again.
Might also work if you want other parts laser cut.
anyway - just putting this here
 - https://www.ponoko.com/materials/stainless-steel?materialTypes=metal

Neon22 committed:
refactor rounded rail

- exact same shape
- simplified code to a railend and roundedrail
- factorised $fn for eventual use in params

Neon22 committed:
refatcor Needlebed

- initial reorg and rename
- primary = build_needle_bed

Neon22 committed:
Needlebed customizer UI

- Add UI so can choose bed scale, num needles etc

Neon22 committed:
Needlebed formatting and comments

- just visual stuff

Neon22 committed:
correct calc for needlebed screws

- call needlebed screws with params so works across all files
- correct calc for screwx position
- comment out redundant left/right screw calcs and actions which led to doubles.

Needlebed screws now match spongebar spacers as needlecount changes.

Neon22 committed:
needlebed paramaterise

- refactor to left and right vars
- no geo change

Neon22 committed:
Add var for fragment resolution to params

- used in rounded rail for now

Neon22 committed:
Spongebar refactor

- used module to build_spongeBar
- added customiser for bed sizes etc
- minor visual changes

Neon22 committed:
refactor connector

- no changes to geo
- formatting and ordering

Neon22 committed:
fragment added to params

Used by connector

Neon22 committed:
refactor backcover

- add build module
- formatting
- add needlbedscrews parameters
- no change to geo

Neon22 committed:
backcover - add customiser

- added customiser interface for printing
- added rounded print option

Neon22 committed:
needlebed assembly refactor

- modified to show the components with on/off controls
- does not propagate param changes yet. more refactoring required for Assemblies...

Neon22 committed:
carriagerest refactor

- comments
- turn on front and back rails - question for later
- simplify clamp screws to two modules

Neon22 committed:
carriagerst refactor

- simplify angle front plate removal
- comments and formatting

Neon22 committed:
carriagerest refactor

- move elements
- modularise build carriage rest
- add customiser
- add section for clamp configuration
- add print orientation option

Neon22 commented:
OK. first pass at most elements of the needlebed. Descriptions on each commit.
- Only reorganised/mangled a couple of functions for efficiency (screw holes mainly)
- Added some customiser UI for printing parts.
- Carriagerest and needlebed are most modified.
- Check out Carriage rest for printing orientation choice in UI.

I will move to the carriage next.

Question about the carriage rest:
 - Its probably easier to print it out if it was printed upside down and without the rails. 
 - Then print rails and screw them on.
 - But I have not printed it so not sure.
 - What do you want me to do ?

Let me know if I have gone too far, or in a direction you do not agree with.
Cheers...

Neon22 committed:
yarnfeeder refactor

- remove stripper plate sections (did not intersect)
- cleaned up difference ops
- added build_yarn_feeder module
- no geo change

Neon22 committed:
yarnfeeder refactor

- formatting
- variables created for some values
- added cyres50,100 to params for circle resolution
- color change because red is also # highlight color

Neon22 committed:
stripperplate refactor

- no geo change
- formatting
- mark unused
- refactor mirror into module
- cylres used for sph,cyl
- colors rationalised
- build module created

Neon22 committed:
add orientation for printing to customiser options

in yarnfeeder and fix in carriage rest.

Neon22 committed:
yarncarriercover refactor

- no geo change
- add module for build
- add variable
- formatting

Neon22 committed:
stripperplate minor adj

changed notes as cutout used by yarncarrier cover but not here in stripperplate. Easier to design here though, as you can see tolerances.

Neon22 committed:
tpointer refactor

- added convexity
- changed color
- formatting

Neon22 committed:
tcam refactor

- added orientation for printing
- color change
- convexity
- prints mirror pairs

Neon22 committed:
tcam fix

oops. got them back up the right way again

Neon22 committed:
clamphead refactor

- formatting
- cylres from params

Neon22 committed:
backplate refactor

- no geo change
- but big change to how sides rounded. Used two rounded negative corners rather than Intersection method.
-

Neon22 committed:
backplate printing orientation

- added orientation param to customiser to flip over for printing

Neon22 committed:
tpointer printing orientation

- added print and rotates
- also dup pointer so assembly looks good

Neon22 committed:
yarn carrier assembly

- added visualisation configurations

Neon22 committed:
needlebed assembly

tweak wording in red

Neon22 committed:
tpointer refactor

- made module to call from assenblies

Neon22 committed:
tcam refactor

- adjusted builds to work with assemblies

Neon22 committed:
needlebed assembly refactor

- reenabled carriage rests and clamps

Neon22 committed:
camplate refactor

- assembly level collation
- added customiser UI

Neon22 committed:
needlebed assembly refactor

- removed adjustments from assembly
- They are in needlbed part

Neon22 commented:
OK. so that's probably it for now.
I have:
 - done a lot of formatting standardisation
 - made many commits so trawling through them may be useful but maybe not...
 - added customisers in the parts and some modules for:
    - printing (orientation changes)
    - selection of gauge and number of needles
    - hiding/showing stuff, if many parts
    - printer tolerances 
- The Assemblies do not have these adjustments - just show/hide so you can visually check clearances
- So its easier to make a very small number of choices and print all the parts you need. (IMHO)

One day maybe a Configurator so all parts can be laid out after making some key decisions.
But for now - following your instructions is how to print the units (but the sizing is now easily controlled).

I think there is a small "error" in yarnFeeder to do with large tolerance values but no real problem.
All the parts seem to me to be exactly the same shapes as they were before I began. but hey - might have made a mistake...

I reworked only about 4 sections where the code got a bit messy. Development is hard. Hopefully my adjustments will make it easier to move forward to the next designs.

Please let me know if this was useful.

Neon22 committed:
Formatting tab removal

- swapped all tabs for spaces.
I use notepad++ for openscad dev and realise that looking at this in a different editor might make the formatting appear weird.

So all tabs to spaces.

Neon22 committed:
renamed Assembly files

Some files renamed to make clearer what each contains

Neon22 commented:
Last two commits cleanup tabs to spaces and rename some assemblies to make it more obvious what's in them.

Neon22 committed:
oops rename

oops. forgot to rename assmblies in here as well

Neon22 committed:
doc support

Added rendered images for doc support

ScarlettSparks commented:
Hi @Neon22, thank you for all your effort on this and sorry for the delay in replying. I've been traveling and mostly offline due to family obligations so I haven't yet been able to sit down with the recent changes you've proposed, but I'll give everything a thorough read as soon as I'm able; unfortunately this might take longer than usual as the next few weeks will continue to be very hectic, but I wanted to pop in and let you know I've seen it and it's on my priorities list!

Neon22 commented:
No probs, appreciate the message. Cheers...

Neon22 commented:
I halted work on this branch and moved to a new branch Screws.
Screws builds on top of this scad_enhancement branch.

ScarlettSparks commented on SCAD/assemblies/assembled_camplate.scad:
I would like to settle on a consistent naming convention/format for these assembly files for clarity (I realize I was a bit inconsistent to begin with so this is also partly a "note to self"). I prefer camelCase in most situations where I have the choice (I find it easier to read and to type than underscores, personally) so that's what I've used throughout the project, but if there is a specific reason you think I should consider snake_case instead, please let me know.

Neon22 replied:
I did not choose for specific reason. It does make my changes more visible to you. Same reason I left in your commented out code in several places - so you could compare. Easy to remove these regions when you're done

ScarlettSparks replied:
Ah, I understand now, thanks for clarifying! I also often leave old code commented out while I'm re-working things, so I can see how this strategy would be useful while the changes are in progress. For what it's worth, though, the git/GitHub diff tool highlights the changes pretty clearly in most cases, so it's not necessary to do this for my sake. It makes reviewing the PR a bit easier for me (and hopefully less work for you, too) if I can assume the changes I see are exactly what you're proposing to merge into main branch. (Or feel free to set a PR as a draft if you've got changes in progress that you want to discuss with me but it's not ready to merge yet.)

Neon22 replied:
OK. will do.
Alas my git-fu is too weak to try to preview how you will see the git diffs before I make a PR. I am using notepad++ for openscad dev as it works quite well and I can find/replace across all files/buffers easily.

Its always easier to work out what the intentions were once ALL the code is refactored but then it needs a second pass :)

ScarlettSparks commented on SCAD/assemblies/assembled_camplate.scad:
```diff
@@ -0,0 +1,48 @@
+    text("Do not print from this file.", size=5);
+    translate([0,-15,0])
+    text("Assembly Visualisation.");
+    }
```
Really like the idea of adding the visual warning to the files which are not meant to be printed. 👍

Neon22 replied:
Showing intent is useful :)

ScarlettSparks commented on SCAD/assemblies/assembled_camplate.scad:
```diff
@@ -0,0 +1,48 @@
+Show_pointers = true;
+Show_spongebar = true;
+Show_screws = true;
+
```
I like the idea of the show/hide toggle for assemblies 👍

ScarlettSparks commented on SCAD/assemblies/toplevel_assembly.scad:
Clarifying that this is the top level seems like a helpful change.

Neon22 replied:
I did that because I got confused about remembering which did what when I was refactoring. Cheers...

ScarlettSparks commented on SCAD/modules/params.scad:
```diff
@@ -13,7 +13,12 @@ I recommend only modifying these constants if you are also
 
+// sphere and cylinder facet counts
 $fn = 50;
+cylres30 = 30;  // roundedrail
```
TBH I didn't document my thinking very well when it comes to these facet counts so I'm not sure at a glance how many of these are different in an *important* way and how many are just based on... vibes. 🤷  Either way I think it's a good idea to consolidate related `$fn` values with a named variable but I'd prefer to use something more descriptive of purpose (e.g. `roundedRailRes = 30`), otherwise I'm not sure there's much advantage over using the number literal. I'll have a closer look at where you've identified these values being used.

Neon22 replied:
I did it this way because I have found this mechanism useful in openscad work. I.e. no confusion that this variable is to be used for anything other than controlling number of facets in spheres and cylinders. I confess I tend to use a smaller number of these than in this project. But there's always time to refactor the refactoring :)

ScarlettSparks commented on SCAD/modules/roundedRail.scad:
```diff
@@ -1,22 +1,25 @@
 include<params.scad>;
 
+module railend(depth, height) {
```
Pulling this into a module seems like a good idea 👍

ScarlettSparks commented on SCAD/modules/roundedRail.scad:
```diff
@@ -1,22 +1,25 @@
+            cube([width-rad*3,depth,height], center = true);
+        translate([rad, 0, height/2 - rad])
+            railend(depth, height);
         translate([width - rad,0,0])
```
Welllllll I dunno what the heck I was doing here. 🙃 The translate-z should match line 10 (or vice-versa?), I think it just *happens* to equal zero with the values I used. Makes a good case for the `railend` module you proposed!

Neon22 replied:
IMHO this happens all the time when developing code as you converge on the right solution while writing the code. No shade casting here...

ScarlettSparks commented on SCAD/modules/roundedRail.scad:
```diff
@@ -1,22 +1,25 @@
 
+module railend(depth, height) {
+    sphere(d=depth, $fn=cylres30);
+    translate([0,0,-height/4])
```
Could you clarify how you arrived at `-height/4` for the z-translation?

Neon22 replied:
sorry - no recollection. I was probably trying to find the right value that meant the result exactly matched your stl file. As this was the first file I refactored, I might well have intended to come back once I worked out what was important. I did not return. Ahh refactoring the refactoring. Exactly what you are doing now :)

Neon22 replied:
OK. Looking at it again.
Could probably have written it:
```
module railend(depth, height) {
    sphere(d=depth, $fn=cylres30);
    translate([0,0,-height/2])
        cylinder(h=height, d=depth, center=true, $fn=cylres30);
}
// and
module roundedRail(width,depth,height) {
    rad = depth/2;
    translate([-width/2,0,0])
    hull() {
        translate([width/2,0,0])
            cube([width-rad*3,depth,height], center = true);
        translate([rad, 0, height/2 - rad])
            railend(depth, height/2);
        translate([width - rad,0,0])
            railend(depth, height/2);
    }
}
```

Neon22 replied:
the height/2 is, of course, because of using `center=true`.
Seems to me half the time `center` is super helpful and half the time you end up adding/subtracting half the value to get it positioned right :)

ScarlettSparks replied:
> Seems to me half the time center is super helpful and half the time you end up adding/subtracting half the value to get it positioned right :)

Yeah, I've also found `center` to be a bit of a mixed bag when it comes to efficiency, but I use it a lot because it makes the object origin explicit and unambiguous. Otherwise I tend to forget which corner things will originate from. 😅 

Thanks for taking another look at the translation math. In general, I agree that my original code here needs a lot of cleanup and clarification (and, as noted in another comment, some debugging 😬); it's *really* convoluted. I'm not confident that the refactoring you've done here maintains the intended logic but that's understandable since the intention was so messy and undocumented to start with. I'm going to revisit this section to make sure I clearly remember/document what the purpose of each of these primitives was, and we can work on making it more concise and robust from there.

ScarlettSparks commented on SCAD/modules/roundedRail.scad:
```diff
@@ -1,22 +1,25 @@
 module roundedRail(width,depth,height) {
-    $fn = 30;
     rad = depth/2;
+    tolerance5 = 0.5;  // tolerance=0.1 * 5 (params)
```
I think this variable goes unused. It looks like it might be meant to replace the literal values on (original) lines 12 and 18 [^1], but maybe got missed as things were moved around. However, I don't follow the source of the math in the inline comment so perhaps I've misunderstood your intent – could you clarify?

[^1]: in which case I agree that it should be a named parameter, especially since I think it has a corresponding translation somewhere on the carriage (and/or I need to reconsider the logic of how this rail's size and position is defined relative to other parts; it got a bit messy over time 🫤 )

Neon22 replied:
Same as above. This was first file I started on and you don't know what you don't know at the beginning. I probably intended to return but never did... Oops...

Neon22 replied:
yeah - did not need that. I think I incrementally adjusted it away and hen forgot to remove the redundant calc.

ScarlettSparks commented on SCAD/parts/backPlate.scad:
```diff
@@ -6,139 +6,184 @@ use<spongeBar.scad>;
 
 module backPlate() {
-    color("magenta")
+    color("Plum")
```
I see you changed the display colours in several places. Is there an advantage or particular reason you selected different ones? (I had no strong attachment to the original colours, I picked things on the fly to help me debug/differentiate parts, so this is just for my own curiosity.)

Neon22 replied:
I find the high saturation colors (especially blue which is very dark) make it hard to see clearly. So I opened the cheatsheet and went to eth color page and picked out lighter colors.

The only intentional color is the inside of screwholes - which I think are all gold. That and maybe the screws themselves - all the same color - I tend to use the colors for grouping like that. I.e. is this hole here for some reason other than being a screwhole.

ScarlettSparks replied:
Ah, that makes a lot of sense, thank you! I'll keep that in mind and try to choose lower saturation colours next time it comes up; I'm sure you're not the only one who finds them easier on the eyes.

ScarlettSparks commented on SCAD/parts/backPlate.scad:
```diff
@@ -6,139 +6,184 @@ use<spongeBar.scad>;
             backRail(width = CAM_PLATE_WIDTH*2 +1, tolerance = -tolerance);
         }
-        
+        // Front slot fits over backrail
```
```suggestion
        // Front slot fits over front rail
```

ScarlettSparks reviewed (commented):
OK, I've finally had a chance to give this a more thorough read – thanks again for your patience! I've made a few in-line comments and questions but I'll also add some general notes here:

👍  Incorporating the `build` functions for each part file seems to improve the code organization/structure a lot and I really like that it would enable toggling visibility in the assemblies!
👍  Adding the visual `don't print` warning to unprintable files seems like a nice touch 
👍  Overall this has highlighted a lot of places where my clarity of intention and inline documentation could be improved, as well as spots where my logic is brittle or convoluted, which is really helpful. 
🚧  I realize I hadn't specified a style-guide (and don't have perfect adherence to this myself) but I would like to maintain a consistent approach to variable and file names to help keep things organized, especially as it's mentioned prominently in the documentation. So far I have used `camelCase` for variables and `SCREAMING_SNAKE` to differentiate constants; most filenames are also lower or camelCase (my stubby lil fingers don't like typing underscores 😅). If there's a practical reason you think a different approach suits a specific proposed change, though, please let me know! 
🚫  Many of the changes you've proposed seem aimed at using the Customizer interface, which is a bit beyond the scope of what I had anticipated for this PR. While I can see the appeal of the Customizer, it looks like supporting it for this project requires some relatively disruptive changes to how params are handled and needs further work to prevent problems; e.g. having values for `gauge` in multiple places breaks needlebed components very easily. I think this is a feature I need to consider separately from any refactoring.
⚪ I noticed that you made some changes to whitespace/formatting but I don't plan to review them in detail at this stage. I'm not too fussy about whitespace style in most cases unless it breaks things or affects readability.

**TL;DR:** I do think there are parts of this PR that I'd like to incorporate but they are entangled with other things, especially the Customizer changes, that aren't the right fit. Rather than asking you to make extensive changes to this PR, I wonder if it would be simpler for me to extract the 👍 suggestions to a new branch/PR with a narrower scope, and [add you as a commit co-author](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) so you retain attribution for your work. What do you think?

ScarlettSparks commented:
> I think there is a small "error" in yarnFeeder to do with large tolerance values but no real problem.

I couldn't spot this, do you recall what line or module you found the error in?

> Question about the carriage rest:
> 
> * Its probably easier to print it out if it was printed upside down and without the rails.
> * Then print rails and screw them on.

I recommend printing it rails-up with a small amount of support material (added in the slicer settings) in the clamp area. I try to design parts to avoid requiring supports where possible, but in this case I'm not sure that it's worth the extra walls, screw hardware, and assembly time that would be required to print the rails separately.  

> BTW - I have an ancient all-metal Corona which used metal sinkers at the end of each row. I used to get stuff done through Ponoko but they changed their tack several years ago and I have not used them since. But I was thinking about what would be involved to make metal parts like the Corona sinkers and looked them up again. Might also work if you want other parts laser cut. anyway - just putting this here
> 
> * https://www.ponoko.com/materials/stainless-steel?materialTypes=metal

Thanks for the tip!

Neon22 commented:
>🚫 Many of the changes you've proposed seem aimed at using the Customizer interface, which is a bit beyond the scope of what I had anticipated for this PR. While I can see the appeal of the Customizer, it looks like supporting it for this project requires some relatively disruptive changes to how params are handled and needs further work to prevent problems; e.g. having values for `gauge` in multiple places breaks needlebed components very easily. I think this is a feature I need to consider separately from any refactoring.

Ahh. actually the Customizer came later.
The reason its like this (sorry I should have indicated this when I submitted) is to do with making the entire set of files work for different needle gauges. The PR as submitted does not go all the way to make this work. Once you've accepted/changed everything to what you want, I had intended to submit a further PR.

This PR would add gauge to several (not many actually) of the modules as an argument. This would mean you could go to a top level assembly (or any of the part/modules) and enter gauge and the entire system would end up with a correctly spaced pieces.

The customizer interface is primarily to help further development - e.g you can see the cutaways around holes for screws in some parts, so you can see if there is enough clearance when adding screws/inserts/etc...

Neon22 commented:
> **TL;DR:** I do think there are parts of this PR that I'd like to incorporate but they are entangled with other things, especially the Customizer changes, that aren't the right fit. Rather than asking you to make extensive changes to this PR, I wonder if it would be simpler for me to extract the 👍 suggestions to a new branch/PR with a narrower scope, and [add you as a commit co-author](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) so you retain attribution for your work. What do you think?

Perfectly happy for you to do this in any way that works for you - my attribution does not have to be noted. I participate to try to help make things run smoother for your work. I am aware that might sound a bit weird but I only want to help you to leap further ahead and I find that when developing a project this complex in one go - that some goals get subverted as we go on by stuff that needs to just be got to work. So I feel a refactoring pass is just me doing a bit of sweeping up after your hard work. I do not need attribution for that. Happy to be invisible.

Neon22 commented:
> > I think there is a small "error" in yarnFeeder to do with large tolerance values but no real problem.
> 
> I couldn't spot this, do you recall what line or module you found the error in?

OK. Found it. Its in tolerance calc.
If you increase the tolerance value the holes move out of the object bounds.
Its not a critical change as tolerance is unlikely to vary by much at all but the holes will become misaligned
 but its fine really.

ScarlettSparks commented:
Hi @Neon22, I am (once again) so sorry to let this sit for so long. Regarding the Customizer changes:

> The PR as submitted does not go all the way to make this work. Once you've accepted/changed everything to what you want, I had intended to submit a further PR.

I see where you're coming from here, but unfortunately accepting this PR with the changes only partially implemented would leave the project in a bit of an unstable state in the meantime. I'm open to revisiting the idea of the Customizer in the future, but it looks like it would require a lot of careful work re-imagining how the project is organized, and at the moment I'd prefer to prioritize missing or requested features that will make the machine more useable.

I won't be able to merge this PR, so I'm going to close the thread, but as I said above, I do think that your suggestions have provided a lot of helpful pointers to areas that need clean-up or clarity and I appreciate your patience and all the time you've contributed. 💖
